cineast-core/src/main/java/org/vitrivr/cineast/core/util/TextDetector_EAST.java

```diff
@@ -0,0 +1,163 @@
+     */
+    public TextDetector_EAST(float confidenceThreshold, float nonMaximumSuppressionThreshold) {
+        this.confThreshold = confidenceThreshold;
+        this.nmsThreshold = nonMaximumSuppressionThreshold;
+        this.outNames = new ArrayList<>(2);
+        outNames.add("feature_fusion/Conv_7/Sigmoid");
+        outNames.add("feature_fusion/concat_3");
+
+    }
+
+    /**
+     * Sets default confidence and non-maximum suppression threshold values
+     * Confidence threshold = 0.5f
+     * Non-maximum suppression threshold = 0.4f
+     */
+    public TextDetector_EAST() {
+        this(0.5f, 0.4f);
+    }
+
+    /**
+     * initialize takes a path to the weights file and reads it
+     * @param modelPath of the weights file
+     * @return returns its own instance
+     */
+    public TextDetector_EAST initialize(String modelPath) {
+       this.model = Dnn.readNetFromTensorflow(modelPath);
+       return this;
+    }
+
+    /**
+     * initialize reads the default weights file
+     * @return returns its own instance
+     */
+    public TextDetector_EAST initialize() {
+        return initialize("resources/TextSpotter/frozen_east_text_detection.pb");
+    }
+
+    private Pair<List<Float>, List<RotatedRect>> decode(Mat scores, Mat geometry) {
+        List<Float> confidences = new ArrayList<>();
+        List<RotatedRect> boxes = new ArrayList<>();
+        int W = geometry.cols();
+        int H = geometry.rows() / 5;
+
+        for (int y = 0; y < H; ++y) {
+            Mat scoresData = scores.row(y);
+            Mat x0Data = geometry.submat(0, H, 0, W).row(y);
+            Mat x1Data = geometry.submat(H, 2 * H, 0, W).row(y);
+            Mat x2Data = geometry.submat(2 * H, 3 * H, 0, W).row(y);
+            Mat x3Data = geometry.submat(3 * H, 4 * H, 0, W).row(y);
+            Mat anglesData = geometry.submat(4 * H, 5 * H, 0, W).row(y);
+
+            for (int x = 0; x < W; ++x) {
+                double score = scoresData.get(0, x)[0];
+                if (score >= this.confThreshold) {
+                    double offsetX = x * 4.0;
+                    double offsetY = y * 4.0;
+                    double angle = anglesData.get(0, x)[0];
+                    double cosA = Math.cos(angle);
+                    double sinA = Math.sin(angle);
+                    double x0 = x0Data.get(0, x)[0];
+                    double x1 = x1Data.get(0, x)[0];
+                    double x2 = x2Data.get(0, x)[0];
+                    double x3 = x3Data.get(0, x)[0];
+                    double h = x0 + x2;
+                    double w = x1 + x3;
+                    Point offset = new Point(offsetX + cosA * x1 + sinA * x2, offsetY - sinA * x1 + cosA * x2);
+                    Point p1 = new Point(-1 * sinA * h + offset.x, -1 * cosA * h + offset.y);
+                    Point p3 = new Point(-1 * cosA * w + offset.x,      sinA * w + offset.y); // original trouble here !
```

Comment:
what does 'original trouble here !' refer to?

Reply:
This comment is removed in the newest version, as it was not meant to stay. For the detector and recognizer implementation, I had initial troubles accessing the right values within the Mat structure, which either gave me out-of-bounds errors, or just erroneous values in a later step of the pipeline.